mqtt/mqtt-broker/tests/common.rs

```diff
@@ -284,11 +287,135 @@ where
     }
 }
 
+lazy_static! {
+    static ref DEFAULT_TIMEOUT: Duration = Duration::from_secs(5);
+}
+
+/// A simple wrapper around TcpStream + PacketCodec to send specific packets
+/// to a broker for more granular integration testing.
+#[derive(Debug)]
+pub struct PacketStream {
+    codec: Framed<TimeoutStream<TcpStream>, PacketCodec>,
+}
+
+impl PacketStream {
+    /// Creates a client and opens TCP connection to the server.
+    /// No MQTT packets are sent at this moment.
+    pub async fn open(server_addr: impl ToSocketAddrs) -> Self {
+        // broker may not be available immediately in the test,
+        // so we'll try to connect for some time.
+        let mut result = TcpStream::connect(&server_addr).await;
+        let start_time = Instant::now();
+        while let Err(_) = result {
+            if start_time.elapsed() > *DEFAULT_TIMEOUT {
+                break;
+            }
+            result = TcpStream::connect(&server_addr).await;
```

Comment:
How long does TcpStream::connect take? should you put like a 200ms sleep here to prevent it from spamming reconnect s?

Reply:
I thought about that, constant retry it doesn't hurt for tests, we just need to timeout at some point. But I can add the delay.